[SPARK-LLAP-133][BRANCH-1.6] Access control for DESC TABLE

merlintang opened the pull request:
## What changes were proposed in this pull request?

Add the access control for describe database.table

**BEFORE**
```
0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_no;
+-----------+------------+----------+--+
| col_name  | data_type  | comment  |
+-----------+------------+----------+--+
+-----------+------------+----------+--+
No rows selected (0.782 seconds)

0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_full;
+-----------+------------+----------+--+
| col_name  | data_type  | comment  |
+-----------+------------+----------+--+
| a         | int        |          |
| b         | int        |          |
+-----------+------------+----------+--+
```

**AFTER**

```
0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_no;
Error: shadehive.org.apache.hive.service.cli.HiveSQLException: Error while compiling statement: FAILED: HiveAccessControlException Permission denied: user [spark] does not have [SELECT] privilege on [spark_ranger_test/t_no/*] (state=,code=0)

0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_full;
+-----------+------------+----------+--+
| col_name  | data_type  | comment  |
+-----------+------------+----------+--+
| a         | int        |          |
| b         | int        |          |
+-----------+------------+----------+--+
2 rows selected (3.18 seconds)

0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_partial;
Error: shadehive.org.apache.hive.service.cli.HiveSQLException: Error while compiling statement: FAILED: HiveAccessControlException Permission denied: user [spark] does not have [SELECT] privilege on [spark_ranger_test/t_partial/*] (state=,code=0)
```

## How was this patch tested?

Pass the test suite.

```
[root@ctr-e133-1493418528701-160107-01-000009 python]# ./spark-ranger-secure-test.py TableTestSuite
......
----------------------------------------------------------------------
Ran 6 tests in 708.943s

OK
```

merlintang commented:
@dongjoon-hyun  can you review this?

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -23,26 +23,24 @@ import java.util.{Map => JMap}
 import java.util.regex.Pattern
 
 import scala.reflect.runtime.{universe => ru}
-
```
This is a violation of ScalaStyle. I think Travis CI will complain about this.

dongjoon-hyun replied:
Hmm. `branch-1.6` seems to be out of order for this.

dongjoon-hyun replied:
Anyway, don't remove this blank.

merlintang replied:
ok, thanks.

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -23,26 +23,24 @@ import java.util.{Map => JMap}
 import scala.reflect.runtime.{universe => ru}
-
 import com.hortonworks.spark.sql.hive.llap.DefaultJDBCWrapper
-
```
here, too.

merlintang replied:
fixed

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -23,26 +23,24 @@ import java.util.{Map => JMap}
-import org.apache.spark.sql.execution.CacheManager
-import org.apache.spark.sql.execution.datasources.LogicalRelation
-import org.apache.spark.sql.execution.datasources.ResolvedDataSource
+import org.apache.spark.sql.execution.{RunnableCommand, _}
```
```
import org.apache.spark.sql.execution._
```

merlintang replied:
fixed

dongjoon-hyun replied:
It's not yet fixed. :)

merlintang replied:
oh, I see. let me fix this.

merlintang replied:
fixed

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+                val nameParts = tables{1}.split("\\.")
+                if (nameParts.length != 2) {
+                  throw new IllegalArgumentException
+                  ("Expected " + s + " to be in the form db.table")
```
indentation?

merlintang replied:
fixed

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+        var tablename = ""
+        val args = desc.argString
+        args.split(",").foreach{
+          s =>
```
```
args.split(",").foreach{ s =>
```

merlintang replied:
sorry about this, I am fixing them now.

merlintang replied:
fixed

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+        args.split(",").foreach{
+          s =>
+            if (s.contains("table"))
+              {
```
```
if (s.contains("table"))
  {
```

merlintang replied:
here too

merlintang replied:
fixed

dongjoon-hyun commented:
Please pass the `DESC` test case and write the result in the PR description.

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+    super.executePlan(plan)
+  }
+
+  private[sql] class QueryExecution(logicalPlan: LogicalPlan)
```
`private`? Do we need to open this to all `sql` module?

merlintang replied:
I think private would be ok, since this is also defined as private in the LlapContext

dongjoon-hyun replied:
? What I mean is the following.
```
- private[sql] class ...
+ private class ...
```

merlintang replied:
oh, I can changed it to  " protected[sql] class QueryExecution". thanks

dongjoon-hyun replied:
What do you mean?

dongjoon-hyun replied:
As I wrote, what I meant is removing the package scope `[sql]`.

merlintang replied:
oh, fixed.

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+              {
+                val tables = s.split("->")
+                val nameParts = tables{1}.split("\\.")
+                if (nameParts.length != 2) {
```
We need to support `DESC table`, too.

merlintang replied:
0: jdbc:hive2://ctr-e133-1493418528701-160107> use spark_ranger_test;
+---------+--+
| result  |
+---------+--+
+---------+--+
No rows selected (1.272 seconds)
0: jdbc:hive2://ctr-e133-1493418528701-160107> desc t_no;
Error: org.apache.spark.sql.AnalysisException: Table not found: t_no; line 1 pos 5 (state=,code=0)

I think this might be the limit of spark 1.6.x. ?

dongjoon-hyun replied:
That is the correct and consistent behavior with Apache Spark and also with SPARK-LLAP 2.x. :)

dongjoon-hyun replied:
t_no is invisible by Ranger rule.

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
+    /**
+     * Access control for Desc Table Function
+     */
+    val checkDesc = executedPlan match {
```
Is there anyone use this `checkDesc`?

merlintang replied:
I left here for someone else could use it in the future. if it is not necessary, I can remove it.

dongjoon-hyun replied:
+1 for removing this.
Generally, that's not a good idea to consider someone else in the future.
Most people (like me) will be confused here and will waste their time in order try to find where this variable used.

merlintang replied:
ok, thanks, let me fix them

merlintang replied:
change to a function. fixed

dongjoon-hyun commented on src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala:
```diff
@@ -124,6 +124,41 @@ class LlapContext(sc: SparkContext,
+            if (s.contains("table")) {
+                val tables = s.split("->")
+                val nameParts = tables{1}.split("\\.")
+                if (nameParts.length != 2) {
```
We need to support `DESC table`, too.

merlintang commented:
ok, I will past the test result to here.

merlintang commented:
yes, t_no is invisible to Ranger Rule.

0: jdbc:hive2://ctr-e133-1493418528701-160107> desc spark_ranger_test.t_no;
Error: shadehive.org.apache.hive.service.cli.HiveSQLException: Error
while compiling statement: FAILED: HiveAccessControlException
Permission denied: user [spark] does not have [SELECT] privilege on
[spark_ranger_test/t_no/*] (state=,code=0)



On Thu, Jul 6, 2017 at 5:33 PM, Dongjoon Hyun <notifications@github.com>
wrote:

> *@dongjoon-hyun* commented on this pull request.
> ------------------------------
>
> In src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala
> <https://github.com/hortonworks-spark/spark-llap/pull/150#discussion_r126048096>
> :
>
> > +
> +    /**
> +     * Access control for Desc Table Function
> +     */
> +    val checkDesc = executedPlan match {
> +      case ExecutedCommand(desc: DescribeCommand) =>
> +        var database = ""
> +        var tablename = ""
> +        val args = desc.argString
> +        args.split(",").foreach{
> +          s =>
> +            if (s.contains("table"))
> +              {
> +                val tables = s.split("->")
> +                val nameParts = tables{1}.split("\\.")
> +                if (nameParts.length != 2) {
>
> t_no is invisible by Ranger rule.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hortonworks-spark/spark-llap/pull/150#discussion_r126048096>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXY-VuaNX9Q8YeTAiWqgruJj9QYNJY8ks5sLXztgaJpZM4OQTVj>
> .
>

dongjoon-hyun commented:
Technically, we should not show that error message. So far, it seems to be Hive bug.
```
HiveAccessControlException
Permission denied: user [spark] does not have [SELECT] privilege on
[spark_ranger_test/t_no/*] 
```
That is a kind of security issue, `Dictionay attack`.

merlintang commented:
yes, thus, this return is correct now, since this is same as the spark-llap
branch 2.1

On Thu, Jul 6, 2017 at 5:46 PM, Dongjoon Hyun <notifications@github.com>
wrote:

> Technically, we should not show that error message. So far, it seems to be
> Hive bug.
>
> HiveAccessControlException
> Permission denied: user [spark] does not have [SELECT] privilege on
> [spark_ranger_test/t_no/*]
>
> That is a kind of security issue, Dictionay attack.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hortonworks-spark/spark-llap/pull/150#issuecomment-313557106>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXY-ebbvPbUtk6WfKiTji0yuWfvSauyks5sLX_OgaJpZM4OQTVj>
> .
>

merlintang commented:
I leave the test result here: 

[root@ctr-e133-1493418528701-160107-01-000009 python]# ./spark-ranger-secure-test.py TableTestSuite
......
----------------------------------------------------------------------
Ran 6 tests in 708.943s

OK

dongjoon-hyun commented:
I finished my second round and updated your PR description.

merlintang commented:
ok, I see.

On Fri, Jul 7, 2017 at 2:15 PM, Dongjoon Hyun <notifications@github.com>
wrote:

> *@dongjoon-hyun* commented on this pull request.
> ------------------------------
>
> In src/main/scala/org/apache/spark/sql/hive/llap/LlapContext.scala
> <https://github.com/hortonworks-spark/spark-llap/pull/150#discussion_r126247215>
> :
>
> > @@ -124,6 +122,44 @@ class LlapContext(sc: SparkContext,
>        Seq.empty
>      }
>    }
> +
> +  override protected[sql] def executePlan(plan: LogicalPlan): super.QueryExecution = {
> +    new this.QueryExecution(plan)
> +    super.executePlan(plan)
> +  }
> +
> +  private[sql] class QueryExecution(logicalPlan: LogicalPlan)
>
> As I wrote, what I meant is removing the package scope [sql].
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hortonworks-spark/spark-llap/pull/150#discussion_r126247215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXY-XL64xuYjEwEP-xH9AA5ArUNMhFjks5sLqAJgaJpZM4OQTVj>
> .
>

merlintang commented:
@dongjoon-hyun  can you review my update? thanks so much for your kind help.

merlintang commented:
Thanks. @dongjoon-hyun
